Update quick start docs (#11140)
PyTorchLightning/pytorch-lightning: docs/source/starter/new-project.rst

.. _new_project:

####################
Lightning in 2 steps
####################
###########
Quick Start
###########

**In this guide we'll show you how to organize your PyTorch code into Lightning in 2 steps.**
*****************
Pytorch Lightning
*****************

Organizing your code with PyTorch Lightning makes your code:
What is Pytorch Lightning
=========================

* Keep all the flexibility (this is all pure PyTorch), but removes a ton of boilerplate
* More readable by decoupling the research code from the engineering
* Easier to reproduce
* Less error-prone by automating most of the training loop and tricky engineering

> * Scalable to any hardware without changing your model

Author: I'll put them back.

Well, if you need one place that can help you with everything you need to train deep learning models, you are at the right place. The team makes sure that all the latest techniques are already integrated and maintained
periodically to make sure you don't need to make any code changes.

> We test the framework across multiple Python/PyTorch versions to ensure its reach. It's not just Keras for PyTorch but much more. Follow along with the guide below to find out why.

Reviewer: The Keras comparison is out of place. What does it mean to be "more than Keras" anyway?

Author: just wanted to specify that it's not just a wrapper around pytorch like keras where you plugin your model and do a `.fit`, but you have control over the loops and what happens inside each step.

.. _new_project:

############
Introduction
############

*************************
What is PyTorch Lightning
*************************

PyTorch has all you need to train your models. But there's a lot more to deep learning more than just attaching some layers together. It provides you the APIs required to build models, datasets etc.

> but when it comes to actual training, there's a lot of boilerplate code involved that you need to write by yourself and if you need to scale your training/inferencing on multiple devices/machines there's another

Reviewer 2: Isn't this a new sentence?

```suggestion
When it comes to actual training, there's a lot of boilerplate code involved that you need to write by yourself. And if you need to scale your training/inferencing on multiple devices/machines there's another
```

set of integrations you might need to do by yourself.

> So here PyTorch Lightning comes into the picture. All you need is some restructuring of your existing code and set certain flags and voila, you are done.

Reviewer 2: ```suggestion
PyTorch Lightning solves this for you. All you need is some restructuring of your existing code and set certain flags and voila, you are done.
```

Now you can train your models on different accelerators like GPU/TPU/IPU, do distributed training accross multiple machines/nodes with no code change use state-of-the-art distributed training mechanisms

> and a lot more. Now all you need is to take care of your research code and leave the engineering to us.

Reviewer 2: ```suggestion
and a lot more.
```

This is saying the same as the sentence below

> Code organization is the core of Lightning. It leaves core training and validation logic to you and automates the rest.

Reviewer 2: ```suggestion
Code organization is the core of Lightning. It leaves the research logic to you and automates the rest.
```

----------

********************
Lightning Philosophy
********************

Organizing your code with Lightning makes your code:

* Keep all the flexibility (this is all pure PyTorch), but removes a ton of boilerplate
* More readable by decoupling the research code from the engineering
* Easier to reproduce
* Less error-prone by automating most of the training loop and tricky engineering
* Scalable to any hardware without changing your model

Lightning is built for:

* Researcher who want to focus on research without worrying about the engineering aspects of it
* ML Engineers who want to built reproducible pipelines
* Data Scientists who want to try out different models for their tasks and build-in ML techniques
* Educators who seek to study and teach Deep Learning with PyTorch

> The team makes sure that all the latest techniques are already integrated and maintained periodically to make sure you don't
> have to worry about any engineering stuff and just focus on research.

Reviewer 2: This is repeating ourselves again

```suggestion
The team makes sure that all the latest techniques are already integrated and well maintained.
```

> It is tested across multiple Python/PyTorch versions to ensure its reach.

Reviewer 2: I don't think this sentence fits here, there's the installation and readme for it.

```suggestion
```

    conda activate my_env
    pip install pytorch-lightning

> Install future patch release from the source

Reviewer: this is the quick start guide, meaning the user will install it fresh, so they anyway get the 1.5.x version. Including this upgrade process here is redundant.

Reviewer 3: what we can do is double branches, meaning the actual `release/1.5.x` will be called `release/stable` and when we start with v1.6, the rename `release/stable` back to `release/1.5.x`

.. code-block:: bash

    pip install git+https://github.com/PytorchLightning/pytorch-lightning.git@release/1.5.x --upgrade

> Install nightly from the source (no guarantees)

Reviewer: not convinced this should be part of the quick start guide, as 99% users will not want to quick start with nightly (why would they). This is distracting, as the user is has to make a choice.

Author: just wanted to introduce a section that tells all possible ways of installation. Pytorch mentions nightly too on their website: https://pytorch.org/

but you are right, maybe if we agree on the introduction page, this can be moved completely out of this one and we can just mention the installation guide for the published one.

Reviewer 3: yeah but their nightly is not bare master but still need to tested and compiled which is not the same as you suggest user to use PL sourcecode :]
